sound/soc/sof/ipc4-topology.c

```diff
@@ -1976,12 +1978,36 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
 	return 0;
 }
 
+static int sof_ipc4_pipeline_send_kcps(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
+				       __s32 adjust)
+{
+	struct sof_ipc4_msg msg = {
+		.primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET) |
+			SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG) |
+			SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST) |
+			SOF_IPC4_MOD_ID(SOF_IPC4_MOD_INIT_BASEFW_MOD_ID) |
+			SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID),
+		.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_REGISTER_KCPS) |
+			SOF_IPC4_MOD_EXT_MSG_FIRST_BLOCK(1) |
+			SOF_IPC4_MOD_EXT_MSG_LAST_BLOCK(1),
```

Comment:
You also need to add to the extension:
`SOF_IPC4_MOD_EXT_MSG_SIZE(sizeof(adjust))`

or you can also use the dedicated function to do a proper LARGE_CONFIG, see #4212.
Since you are always going to send a single u32, it might be justified to have it open coded, but in general I prefer to use the function which exists for the purpose.

Reply:
oh... `sof_ipc4_set_get_data()` is a kind of a monster... In general - yes, I'd also rather use a proper abstraction, so... But well, I still think it's better to use the accessor. Thanks!

Reply:
Yes, it is a monster, but it adds the benefit that we use common code and we can also dump the payload as hex in one place.
It is good to have these handled in one path to help debugging things.